src/services/codefixes/inferFromUsage.ts

```diff
@@ -0,0 +1,655 @@
+                        for (const callContext of callContexts) {
+                            if (callContext.argumentTypes.length > parameterIndex) {
+                                if (isRestParameter) {
+                                    types = concatenate(types, map(callContext.argumentTypes.slice(parameterIndex), a => checker.getBaseTypeOfLiteralType(a)));
+                                }
+                                else {
+                                    types.push(checker.getBaseTypeOfLiteralType(callContext.argumentTypes[parameterIndex]));
+                                }
+                            }
+                        }
+                        if (types.length) {
+                            const type = checker.getWidenedType(checker.getUnionType(types, /*subtypeReduction*/ true));
+                            paramTypes[parameterIndex] = isRestParameter ? checker.createArrayType(type) : type;
+                        }
+                    }
+                    return paramTypes;
+                }
+            }
+            return undefined;
+        }
+
+
+        function getTypeFromUsageContext(usageContext: UsageContext, checker: TypeChecker): Type | undefined {
+            if (usageContext.isNumberOrString && !usageContext.isNumber && !usageContext.isString) {
+                return checker.getUnionType([checker.getNumberType(), checker.getStringType()]);
+            }
+            else if (usageContext.isNumber) {
+                return checker.getNumberType();
+            }
+            else if (usageContext.isString) {
+                return checker.getStringType();
+            }
+            else if (usageContext.candidateTypes) {
+                return checker.getWidenedType(checker.getUnionType(map(usageContext.candidateTypes, t => checker.getBaseTypeOfLiteralType(t)), /*subtypeReduction*/ true));
+            }
+            else if (usageContext.properties && hasCallContext(usageContext.properties.get("then" as __String))) {
+                const paramType = getParameterTypeFromCallContexts(0, usageContext.properties.get("then" as __String).callContexts, /*isRestParameter*/ false, checker);
+                const types = paramType.getCallSignatures().map(c => c.getReturnType());
+                return checker.createPromiseType(types.length ? checker.getUnionType(types, /*subtypeReduction*/ true) : checker.getAnyType());
+            }
+            else if (usageContext.properties && hasCallContext(usageContext.properties.get("push" as __String))) {
+                return checker.createArrayType(getParameterTypeFromCallContexts(0, usageContext.properties.get("push" as __String).callContexts, /*isRestParameter*/ false, checker));
+            }
+            else if (usageContext.properties || usageContext.callContexts || usageContext.constructContexts || usageContext.numberIndexContext || usageContext.stringIndexContext) {

```

Comment:
what happens if you combine object types made up of the synthetic properties (et al) with the candidate types? Are the candidate types strictly better than the synthetic object type?

Reply:
I would think so.. a user-defined type seems more like what you would put on a function, e.g. `Node` or `Symbol`, and not `{kind: SyntaxKind}` or `{flags: SymbolFlags}` to use an example of our code base.